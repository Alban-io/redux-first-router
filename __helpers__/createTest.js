import { applyMiddleware, createStore, combineReducers } from 'redux'
import { createRouter } from '../src'

export default async (routesMap, actions, options = {}) => {
  if (actions && !Array.isArray(actions)) {
    options = actions
    actions = undefined
  }

  if (!actions) {
    actions = Object.keys(routesMap).filter(type => !/FIRST|REDIRECTED/.test(type))
  }

  expect(options).toMatchSnapshot('options')

  const routes = {
    FIRST: '/first',
    NEVER_USED_PATHLESS: { // insures pathless routes can co-exist with regular routes
      thunk: jest.fn()
    },
    REDIRECTED: {
      path: '/redirected',
      onComplete: jest.fn(() => 'redirect_complete')
    },
    ...routesMap
  }

  const path = typeof actions[0] === 'string' && actions[0].charAt(0) === '/'
    ? actions.shift()
    : '/first'

  options.initialEntries = [path]
  options.extra = { arg: 'extra-arg' }

  const { middleware, reducer, firstRoute, rudy } = createRouter(
    routes,
    options
  )

  const title = (state, action = {}) => {
    return action.payload
      ? action.type + '_' + JSON.stringify(action.payload)
      : action.type
  }
  const rootReducer = combineReducers({ title, location: reducer })
  const enhancer = applyMiddleware(middleware)
  const store = createStore(rootReducer, enhancer)

  if (routesMap.FIRST) {
    expect(store.getState()).toMatchSnapshot('initial_state')
  }

  const firstAction = firstRoute()
  const res = await store.dispatch(firstAction)

  if (routesMap.FIRST) {
    expect(firstAction).toMatchSnapshot('first_action')
    expect(res).toMatchSnapshot('first_response')
    expect(store.getState()).toMatchSnapshot('first_state')
    expect(rudy.history.entries).toMatchSnapshot('first_history_entries')
    expect(rudy.history.index).toMatchSnapshot('first_history_index')
    expect(document.title).toMatchSnapshot('first_title')
  }

  for (const item of actions) {
    let res
    let postfix

    if (typeof item === 'string' && item.charAt(0) === '/') {
      res = await rudy.history.push(item)
      postfix = item
    }
    else {
      const act = typeof item === 'object' ? item : { type: item }
      const { type, params, query, hash, state, snipes } = act
      const action = { type }

      if (params) action.params = params
      if (query) action.query = query
      if (hash) action.hash = hash
      if (state) action.state = state

      res = await store.dispatch(action)
      postfix = '_' + type

      expect(action).toMatchSnapshot('action' + postfix)
    }

    expect(res).toMatchSnapshot('response' + postfix)
    expect(store.getState()).toMatchSnapshot('state' + postfix)
    expect(rudy.history.entries).toMatchSnapshot('history_entries' + postfix)
    expect(rudy.history.index).toMatchSnapshot('history_index' + postfix)
    expect(document.title).toMatchSnapshot('title' + postfix)
  }

  const types = Object.keys(routesMap)

  for (const type of types) {
    const route = routes[type]
    const postfix = '_' + type

    if (route.beforeLeave) {
      expect(route.beforeLeave.mock.calls.length).toMatchSnapshot('beforeLeave' + postfix)
    }

    if (route.beforeEnter && route.beforeEnter.mock) { // allow redirect shortcut route options to work
      expect(route.beforeEnter.mock.calls.length).toMatchSnapshot('beforeEnter' + postfix)
    }

    if (route.onLeave) {
      expect(route.onLeave.mock.calls.length).toMatchSnapshot('onLeave' + postfix)
    }

    if (route.onEnter) {
      expect(route.onEnter.mock.calls.length).toMatchSnapshot('onEnter' + postfix)
    }

    if (route.thunk) {
      expect(route.thunk.mock.calls.length).toMatchSnapshot('thunk' + postfix)
    }

    if (route.onComplete) {
      expect(route.onComplete.mock.calls.length).toMatchSnapshot('onComplete' + postfix)
    }

    if (route.onError) {
      expect(route.onError.mock.calls.length).toMatchSnapshot('onError' + postfix)
    }
  }

  if (options.beforeLeave) {
    expect(options.beforeLeave.mock.calls.length).toMatchSnapshot('beforeLeave_options')
  }

  if (options.beforeEnter) {
    expect(options.beforeEnter.mock.calls.length).toMatchSnapshot('beforeEnter_options')
  }

  if (options.onLeave) {
    expect(options.onLeave.mock.calls.length).toMatchSnapshot('onLeave_options')
  }

  if (options.onEnter) {
    expect(options.onEnter.mock.calls.length).toMatchSnapshot('onEnter_options')
  }

  if (options.thunk) {
    expect(options.thunk.mock.calls.length).toMatchSnapshot('thunk_options')
  }

  if (options.onComplete) {
    expect(options.onComplete.mock.calls.length).toMatchSnapshot('onComplete_options')
  }

  if (options.onError) {
    expect(options.onError.mock.calls.length).toMatchSnapshot('onError_options')
  }

  return {
    store,
    firstRoute,
    history: rudy.history,
    routes,
    location: () => store.getState().location
  }
}
