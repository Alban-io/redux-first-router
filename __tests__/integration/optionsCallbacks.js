import createTest from '../../__helpers__/createTest'

test('all options callbacks are called', async () => {
  await createTest({
    SECOND: {
      path: '/second'
    }
  }, {
    beforeLeave: jest.fn(),
    beforeEnter: jest.fn(),
    onLeave: jest.fn(),
    onEnter: jest.fn(),
    thunk: jest.fn(),
    onComplete: jest.fn(({ action }) => {
      if (action.type !== 'SECOND') return
      throw new Error('test-error')
    }),
    onError: jest.fn()
  })
})
