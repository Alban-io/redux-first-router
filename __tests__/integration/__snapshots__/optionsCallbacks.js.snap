// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action_SECOND 1`] = `
Object {
  "type": "SECOND",
}
`;

exports[`action_SECOND 2`] = `
Object {
  "type": "SECOND",
}
`;

exports[`action_THIRD 1`] = `
Object {
  "type": "THIRD",
}
`;

exports[`beforeEnter_options 1`] = `2`;

exports[`beforeEnter_options 2`] = `2`;

exports[`beforeEnter_options 3`] = `3`;

exports[`beforeLeave_options 1`] = `1`;

exports[`beforeLeave_options 2`] = `1`;

exports[`history_entries_SECOND 1`] = `
Array [
  Object {
    "basename": "",
    "hash": "",
    "key": "345678",
    "pathname": "/first",
    "search": "",
    "state": Object {},
    "url": "/first",
  },
  Object {
    "basename": "",
    "hash": "",
    "key": "345678",
    "pathname": "/second",
    "search": "",
    "state": Object {},
    "url": "/second",
  },
]
`;

exports[`history_entries_SECOND 2`] = `
Array [
  Object {
    "basename": "",
    "hash": "",
    "key": "345678",
    "pathname": "/first",
    "search": "",
    "state": Object {},
    "url": "/first",
  },
  Object {
    "basename": "",
    "hash": "",
    "key": "345678",
    "pathname": "/second",
    "search": "",
    "state": Object {},
    "url": "/second",
  },
]
`;

exports[`history_entries_THIRD 1`] = `
Array [
  Object {
    "basename": "",
    "hash": "",
    "key": "345678",
    "pathname": "/first",
    "search": "",
    "state": Object {},
    "url": "/first",
  },
  Object {
    "basename": "",
    "hash": "",
    "key": "345678",
    "pathname": "/redirected",
    "search": "",
    "state": Object {},
    "url": "/redirected",
  },
]
`;

exports[`history_index_SECOND 1`] = `1`;

exports[`history_index_SECOND 2`] = `1`;

exports[`history_index_THIRD 1`] = `1`;

exports[`onComplete_options 1`] = `2`;

exports[`onComplete_options 2`] = `2`;

exports[`onEnter_options 1`] = `2`;

exports[`onEnter_options 2`] = `2`;

exports[`onError_options 1`] = `1`;

exports[`onError_options 2`] = `1`;

exports[`onLeave_options 1`] = `1`;

exports[`onLeave_options 2`] = `1`;

exports[`options 1`] = `
Object {
  "beforeEnter": [Function],
  "beforeLeave": [Function],
  "onComplete": [Function],
  "onEnter": [Function],
  "onError": [Function],
  "onLeave": [Function],
  "thunk": [Function],
}
`;

exports[`options 2`] = `
Object {
  "beforeEnter": [Function],
  "beforeLeave": [Function],
  "onComplete": [Function],
  "onEnter": [Function],
  "onError": [Function],
  "onLeave": [Function],
  "thunk": [Function],
}
`;

exports[`options 3`] = `
Object {
  "beforeEnter": [Function],
  "thunk": [Function],
}
`;

exports[`response_SECOND 1`] = `undefined`;

exports[`response_SECOND 2`] = `undefined`;

exports[`response_THIRD 1`] = `
Object {
  "payload": "redirect_complete",
  "type": "REDIRECTED_COMPLETE",
}
`;

exports[`state_SECOND 1`] = `
Object {
  "location": Object {
    "basename": "",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/second",
        "search": "",
        "state": Object {},
        "url": "/second",
      },
    ],
    "hasSSR": false,
    "hash": "",
    "index": 1,
    "kind": "push",
    "length": 2,
    "params": Object {},
    "pathname": "/second",
    "prev": Object {
      "basename": "",
      "entries": Array [
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
      ],
      "hasSSR": false,
      "hash": "",
      "index": 0,
      "kind": "load",
      "length": 1,
      "params": Object {},
      "pathname": "/first",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 200,
      "type": "FIRST",
      "url": "/first",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "SECOND",
    "url": "/second",
  },
  "title": "SECOND",
}
`;

exports[`state_SECOND 2`] = `
Object {
  "location": Object {
    "basename": "",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/second",
        "search": "",
        "state": Object {},
        "url": "/second",
      },
    ],
    "hasSSR": false,
    "hash": "",
    "index": 1,
    "kind": "push",
    "length": 2,
    "params": Object {},
    "pathname": "/second",
    "prev": Object {
      "basename": "",
      "entries": Array [
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
      ],
      "hasSSR": false,
      "hash": "",
      "index": 0,
      "kind": "load",
      "length": 1,
      "params": Object {},
      "pathname": "/first",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 200,
      "type": "FIRST",
      "url": "/first",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "SECOND",
    "url": "/second",
  },
  "title": "SECOND",
}
`;

exports[`state_THIRD 1`] = `
Object {
  "location": Object {
    "basename": "",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/redirected",
        "search": "",
        "state": Object {},
        "url": "/redirected",
      },
    ],
    "hasSSR": false,
    "hash": "",
    "index": 1,
    "kind": "redirect",
    "length": 2,
    "params": Object {},
    "pathname": "/redirected",
    "prev": Object {
      "basename": "",
      "entries": Array [
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/third",
          "search": "",
          "state": Object {},
          "url": "/third",
        },
      ],
      "hash": "",
      "index": 1,
      "kind": "push",
      "length": 2,
      "params": Object {},
      "pathname": "/third",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 200,
      "type": "THIRD",
      "url": "/third",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 302,
    "type": "REDIRECTED",
    "url": "/redirected",
  },
  "title": "REDIRECTED_COMPLETE_\\"redirect_complete\\"",
}
`;

exports[`thunk_options 1`] = `2`;

exports[`thunk_options 2`] = `2`;

exports[`thunk_options 3`] = `2`;

exports[`title_SECOND 1`] = `"SECOND"`;

exports[`title_SECOND 2`] = `"SECOND"`;

exports[`title_THIRD 1`] = `"REDIRECTED"`;
